### PR TITLE
Implement Ledger support for Evmos CLI

### PR DESCRIPTION
**Pull request by austinchandra**
## Description

Implements the changes required in the Cosmos SDK to support Ledger integration with the Evmos CLI. 

### Context
- Unlike on Cosmos, Ledger compatibility on Evmos uses the Ethereum app, and derives addresses using the Ethereum algorithm
- Cosmos payloads for Evmos on Ledger are signed using EIP-712 TypedData, the signatures of which are included in a transaction extension (rather than in the transaction body)
- Evmos-Ledger-Go is initialized with a decoder, receives SignDoc bytes (of either Amino or Protobuf form), parses these bytes using the decoder, and creates EIP-712 payloads which are then signed

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: crypto/hd/hdpath.go [hunk 1]
```diff
@@ -150,6 +150,22 @@ func (p BIP44Params) DerivationPath() []uint32 {
 	}
 }
 
+// Return params in an array formatted for Ledger
+func (p BIP44Params) LedgerDerivationPath() []uint32 {
```
Comment (fedekunze) [inline crypto/hd/hdpath.go:154]: this change can be upstreamed

File: crypto/keyring/keyring.go [hunk 2]
```diff
@@ -151,6 +151,8 @@ type Options struct {
 	SupportedAlgos SigningAlgoList
 	// supported signing algorithms for Ledger
 	SupportedAlgosLedger SigningAlgoList
+	// define Ledger Derivation function
+	LedgerDerivation func() (ledger.SECP256K1, error)
```
Comment (fedekunze) [inline crypto/keyring/keyring.go:155]: ditto

File: crypto/ledger/ledger_real.go [hunk 3]
```diff
@@ -1,19 +1,15 @@
...
 	discoverLedger = func() (SECP256K1, error) {
-		device, err := ledger.FindLedgerCosmosUserApp()
-		if err != nil {
-			return nil, err
-		}
-
-		return device, nil
+		return nil, errors.New("Ledger discovery must be passed in and is not implemented here")
 	}
 }
```
Comment (fedekunze) [inline crypto/ledger/ledger_real.go:15]: I feel we can remove this and refactor the current the upstream SDK to use `SetDiscoverLedger` instead

File: crypto/ledger/ledger_secp256k1.go [hunk 4]
```diff
@@ -48,6 +47,10 @@ type (
 	}
 )
 
+func SetDiscoverLedger(fn discoverLedgerFn) {
```
Comment (fedekunze) [inline crypto/ledger/ledger_secp256k1.go:50]: nit, add comment

File: crypto/ledger/ledger_secp256k1.go [hunk 5]
```diff
@@ -235,11 +229,13 @@ func sign(device SECP256K1, pkl PrivKeyLedgerSecp256k1, msg []byte) ([]byte, err
 // since this involves IO, it may return an error, which is not exposed
 // in the PubKey interface, so this function allows better error handling
 func getPubKeyUnsafe(device SECP256K1, path hd.BIP44Params) (types.PubKey, error) {
-	publicKey, err := device.GetPublicKeySECP256K1(path.DerivationPath())
+	publicKey, err := device.GetPublicKeySECP256K1(path.LedgerDerivationPath())
 	if err != nil {
-		return nil, fmt.Errorf("please open Cosmos app on the Ledger device - error: %v", err)
+		return nil, fmt.Errorf("please open the Ethereum app on the Ledger device - error: %v", err)
```
Comment (fedekunze) [inline crypto/ledger/ledger_secp256k1.go:234]: the name should be a function of the device (i.e `device.AppName()`)

File: crypto/ledger/ledger_secp256k1.go [hunk 6]
```diff
@@ -235,11 +229,13 @@ func sign(device SECP256K1, pkl PrivKeyLedgerSecp256k1, msg []byte) ([]byte, err
...
 	}
 
+	publicKey = append([]byte{byte(4)}, publicKey...)
```
Comment (fedekunze) [inline crypto/ledger/ledger_secp256k1.go:237]: why this change? can you add a comment?

File: crypto/ledger/ledger_secp256k1.go [hunk 7]
```diff
@@ -259,11 +255,13 @@ func getPubKeyUnsafe(device SECP256K1, path hd.BIP44Params) (types.PubKey, error
 // Since this involves IO, it may return an error, which is not exposed
 // in the PubKey interface, so this function allows better error handling.
 func getPubKeyAddrSafe(device SECP256K1, path hd.BIP44Params, hrp string) (types.PubKey, string, error) {
-	publicKey, addr, err := device.GetAddressPubKeySECP256K1(path.DerivationPath(), hrp)
+	publicKey, addr, err := device.GetAddressPubKeySECP256K1(path.LedgerDerivationPath(), hrp)
 	if err != nil {
 		return nil, "", fmt.Errorf("%w: address rejected for path %s", err, path.String())
 	}
 
+	publicKey = append([]byte{byte(4)}, publicKey...)
```
Comment (fedekunze) [inline crypto/ledger/ledger_secp256k1.go:263]: ditto

File: crypto/ledger/ledger_secp256k1.go (hunk 4 above)
Comment (fedekunze) [inline crypto/ledger/ledger_secp256k1.go:50]: So I think the SDK will want to keep the `crypto/ledger/ledger_secp256k1.go` file as is and have app chains define their own custom functionality to support the desired application (eg: Ethereum)

File: crypto/keyring/keyring.go [hunk 8]
```diff
@@ -211,6 +213,10 @@ func newKeystore(kr keyring.Keyring, opts ...Option) keystore {
 		optionFn(&options)
 	}
 
+	if options.LedgerDerivation != nil {
+		ledger.SetDiscoverLedger(options.LedgerDerivation)
+	}
```
Comment (fedekunze) [inline crypto/keyring/keyring.go:218]: ditto, this can be upstreamed

File: client/tx/tx.go [hunk 9]
```diff
@@ -396,6 +399,54 @@ func Sign(txf Factory, name string, txBuilder client.TxBuilder, overwriteSig boo
...
+			Sequence: txf.Sequence(),
+		}
+
+		err = txBuilder.SetSignatures(sig)
+		if err != nil {
+			return fmt.Errorf("Unable to set signatures on payload: %v\n", err)
+		}
+
+		return nil
+	}
```
Comment (fedekunze) [inline client/tx/tx.go:448]: How can we abstract this functionality in a way that the SDK can support it natively? Ideas:

- Create a new function (or interface) field in the `client.Context` or the `Factory` struct and call this function

**Review by fedekunze — CHANGES_REQUESTED** [review]
Thanks @austinchandra, I think some changes can be upstreamed, while we can abstract some of the logic by implementing interfaces. This way apps like Evmos / Ethermint can define their own logic for the ledger device / app.

 See comments.

@alexanderbez can you review this too?

File: client/tx/tx.go [hunk 10]
```diff
@@ -396,6 +399,54 @@ func Sign(txf Factory, name string, txBuilder client.TxBuilder, overwriteSig boo
 		return err
 	}
 
+	// Modify Ledger signature to use EIP712 extension
+	info, err := txf.Keybase().Key(name)
+	if err != nil {
+		return fmt.Errorf("Error retrieving key type: %v\n", err)
```
Comment (alexanderbez) [inline client/tx/tx.go:405]:
```suggestion
		return fmt.Errorf("Error retrieving key type: %w", err)
```

File: client/tx/tx.go [hunk 11]
```diff
@@ -396,6 +399,54 @@ func Sign(txf Factory, name string, txBuilder client.TxBuilder, overwriteSig boo
...
+	}
+
+	if info.GetType().String() == "ledger" {
```
Comment (alexanderbez) [inline client/tx/tx.go:408]: We should use constants for these (if they dont already exist)

File: client/tx/tx.go [hunk 12]
```diff
@@ -396,6 +399,54 @@ func Sign(txf Factory, name string, txBuilder client.TxBuilder, overwriteSig boo
...
+
+		// Parse Chain ID as UInt
+		chainID, err := etherminttypes.ParseChainID(txf.chainID)
+		if err != nil {
+			return fmt.Errorf("Error parsing chain id: %v\n", err)
+		}
+
+		// Add ExtensionOptionsWeb3Tx extension
+		var option *codectypes.Any
+		option, err = codectypes.NewAnyWithValue(&etherminttypes.ExtensionOptionsWeb3Tx{
```
Comment (alexanderbez) [inline client/tx/tx.go:423]: The `err` isn't checked here.

File: crypto/hd/hdpath.go [hunk 13]
```diff
@@ -150,6 +150,22 @@ func (p BIP44Params) DerivationPath() []uint32 {
...
+	change := uint32(0)
```
Comment (alexanderbez) [inline crypto/hd/hdpath.go:155]:
```suggestion
	var change uint32
```
